src/test/Settings/CredentialsStorage.test.tsx

```diff
@@ -0,0 +1,98 @@
+ * one is included with the Software (each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+import * as React from 'react';
+import renderer, { act } from 'react-test-renderer';
+import { render, cleanup, screen, waitFor, getByText } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+
+import { CredentialsStorage } from '@app/Settings/CredentialsStorage';
+
+import { getFromLocalStorage, saveToLocalStorage } from '@app/utils/LocalStorage';
+
+jest.mock('@app/utils/LocalStorage', () => {
+  const map = new Map<any, any>();
+  return {
+    getFromLocalStorage: jest.fn((key: any, defaultValue: any): any => {
+      if (map.has(key)) {
+        return map.get(key);
+      }
+      return defaultValue;
+    }),
+
+    saveToLocalStorage: jest.fn((key: any, value: any): any => {
+      map.set(key, value);
+    }),
+  };
+});
+
+const storageKey = 'JMX_CREDENTIAL_LOCATION';
+const sessionStorageValue = 'Session (Browser Memory)';
+const backendStorageValue = 'Backend';
+
+describe('<CredentialsStorage/>', () => {
+  afterEach(cleanup);
+
+  it('renders correctly', async () => {
+    let tree;
+    await act(async () => {
+      tree = renderer.create(React.createElement(CredentialsStorage.content, null));
+    });
+    expect(tree.toJSON()).toMatchSnapshot();
+  });
+
+  it('sets value to local storage when dropdown is clicked', async () => {
+    render(React.createElement(CredentialsStorage.content, null));
+
+    expect(getFromLocalStorage).toHaveBeenCalledTimes(1);
+    expect(saveToLocalStorage).toHaveBeenCalledTimes(1);
+    expect(saveToLocalStorage).lastCalledWith(storageKey, sessionStorageValue);
+
+    userEvent.click(screen.getByRole('button'));
+
+    await screen.findByRole('listbox');
+    const ul = screen.getByRole('listbox');
+    const session = getByText(ul, sessionStorageValue);
```

Comment:
Probably, use `screen.getByText` for consistency?

Reply:
I'm querying for children under the `<ul>` because it seems that the Patternfly Select component actually makes a few different child elements with that same inner text:

```
$ yarn test src/test/Settings/CredentialsStorage.test.tsx 
yarn run v1.22.17
$ jest --maxWorkers=50% --coverage=true src/test/Settings/CredentialsStorage.test.tsx
 FAIL  src/test/Settings/CredentialsStorage.test.tsx
  <CredentialsStorage/>
    ✓ renders correctly (19 ms)
    ✓ defaults to Session storage (62 ms)
    ✕ sets value to local storage when dropdown is clicked (51 ms)

  ● <CredentialsStorage/> › sets value to local storage when dropdown is clicked

    TestingLibraryElementError: Found multiple elements with the text: Session (Browser Memory)

    Here are the matching elements:

    Ignored nodes: comments, <script />, <style />
    <span
      class="pf-c-select__toggle-text"
    >
      Session (Browser Memory)
    </span>

    Ignored nodes: comments, <script />, <style />
    <span
      class="pf-c-select__menu-item-main"
    >
      Session (Browser Memory)
      <span
        class="pf-c-select__menu-item-icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          />
        </svg>
      </span>
    </span>

    (If this is intentional, then use the `*AllBy*` variant of the query (like `queryAllByText`, `getAllByText`, or `findAllByText`)).

    Ignored nodes: comments, <script />, <style />
    <body>
      <div>
        <div
          class="pf-c-select pf-m-expanded"
          data-ouia-component-id="OUIA-Generated-Select-single-3"
          data-ouia-component-type="PF4/Select"
          data-ouia-safe="true"
        >
          <button
            aria-expanded="true"
            aria-haspopup="listbox"
            aria-label="Options menu"
            aria-labelledby=" pf-select-toggle-id-3"
            class="pf-c-select__toggle"
            id="pf-select-toggle-id-3"
            type="button"
          >
            <div
              class="pf-c-select__toggle-wrapper"
            >
              <span
                class="pf-c-select__toggle-text"
              >
                Session (Browser Memory)
              </span>
            </div>
            <span
              class="pf-c-select__toggle-arrow"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                />
              </svg>
            </span>
          </button>
          <ul
            aria-label=""
            class="pf-c-select__menu"
            role="listbox"
          >
            <li
              class="pf-c-select__menu-wrapper"
              id="select-option-1664471309403h7v5xg5ychb"
              role="presentation"
            >
              <button
                aria-selected="true"
                class="pf-c-select__menu-item pf-m-selected pf-m-description"
                role="option"
                type="button"
              >
                <span
                  class="pf-c-select__menu-item-main"
                >
                  Session (Browser Memory)
                  <span
                    class="pf-c-select__menu-item-icon"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      style="vertical-align: -0.125em;"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
                      />
                    </svg>
                  </span>
                </span>
                <span
                  class="pf-c-select__menu-item-description"
                >
                  Keep credentials in browser memory for the current session only. When you close this browser tab the credentials will be forgotten.
                </span>
              </button>
            </li>
            <li
              class="pf-c-select__menu-wrapper"
              id="select-option-1664471309405o8wzrm48rcs"
              role="presentation"
            >
              <button
                class="pf-c-select__menu-item pf-m-description"
                role="option"
                type="button"
              >
                <span
                  class="pf-c-select__menu-item-main"
                >
                  Backend
                </span>
                <span
                  class="pf-c-select__menu-item-description"
                >
                  Keep credentials in encrypted Cryostat backend storage. These credentials will be available to other users and will be used for Automated Rules.
                </span>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </body>

       97 |
       98 |     const ul = await screen.findByRole('listbox');
    >  99 |     const session = screen.getByText(sessionStorageValue);
          |                            ^
      100 |     const backend = getByText(ul, backendStorageValue);
      101 |     userEvent.click(backend);
      102 |     await waitFor(() => expect(session).not.toBeVisible());

      at Object.getElementError (node_modules/@testing-library/dom/dist/config.js:38:19)
      at getElementError (node_modules/@testing-library/dom/dist/query-helpers.js:25:35)
      at getMultipleElementsFoundError (node_modules/@testing-library/dom/dist/query-helpers.js:29:10)
      at node_modules/@testing-library/dom/dist/query-helpers.js:66:13
      at node_modules/@testing-library/dom/dist/query-helpers.js:111:19
      at src/test/Settings/CredentialsStorage.test.tsx:99:28
      at step (src/test/Settings/CredentialsStorage.test.tsx:52:23)
      at Object.next (src/test/Settings/CredentialsStorage.test.tsx:33:53)
      at fulfilled (src/test/Settings/CredentialsStorage.test.tsx:24:58)
```

Reply:
Ahh wow right! Then, maybe we can use:

`within(ul).getByText("sessionStorageValue")`

Just a bit clearer what is being done :D